File: Leetcode 424 - longest-repeating-character-replacement/424.longest-repeating-character-replacement.cpp
/*
 * @lc app=leetcode id=424 lang=cpp
 *
 * [424] Longest Repeating Character Replacement
 */

// @lc code=start
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  int characterReplacement(string s, int k) {
    vector<int> count(26, 0);
    int left = 0, maxCount = 0, result = 0;

    for (int right = 0; right < s.size(); right++) {
      count[s[right] - 'A']++;
      maxCount = max(maxCount, count[s[right] - 'A']);

      if ((right - left + 1) - maxCount > k) {
        count[s[left] - 'A']--;
        left++;
      }

      result = max(result, right - left + 1);
    }

    return result;
  }
};

int main() {
  Solution sol;

  string s = "ABAB";
  int k = 2;
  cout << sol.characterReplacement(s, k) << endl; // 4

  s = "AABABBA";
  k = 1;
  cout << sol.characterReplacement(s, k) << endl; // 4
}
// @lc code=end

File: Leetcode 424 - longest-repeating-character-replacement/424.longest-repeating-character-replacement.java
/*
 * @lc app=leetcode id=424 lang=java
 *
 * [424] Longest Repeating Character Replacement
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {

  public int characterReplacement(String s, int k) {
    int left = 0, maxCount = 0, result = 0;

    Map<Character, Integer> letters = new HashMap<>();

    for (int right = 0; right < s.length(); right++) {
      letters.put(
        s.charAt(right),
        letters.getOrDefault(s.charAt(right), 0) + 1
      );
      maxCount = Math.max(maxCount, letters.get(s.charAt(right)));

      while ((right - left + 1) - maxCount > k) {
        letters.put(s.charAt(left), letters.get(s.charAt(left)) - 1);
        left++;
      }

      result = Math.max(result, right - left + 1);
    }

    return result;
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    System.out.println(solution.characterReplacement("ABAB", 2)); // 4
    System.out.println(solution.characterReplacement("AABABBA", 1)); // 4
    System.out.println(solution.characterReplacement("AABBBBA", 2)); // 5
  }
}
// @lc code=end
